server/src/test/java/org/elasticsearch/transport/InboundHandlerTests.java

```diff
@@ -313,30 +313,45 @@ public Version minimumCompatibilityVersion() {
         ByteBuf fullResponse = (ByteBuf) embeddedChannel.outboundMessages().poll();
         BytesReference fullResponseBytes = Netty4Utils.toBytesReference(fullResponse);
         BytesReference responseContent = fullResponseBytes.slice(responseHeaderSize + 2, fullResponseBytes.length() - responseHeaderSize - 2);
-        Header responseHeader = InboundDecoder.readHeader(remoteVersion, fullResponseBytes.length(), fullResponseBytes);
+        Header responseHeader = InboundDecoder.readHeader(fullResponseBytes.length(), fullResponseBytes);
         InboundMessage responseMessage = new InboundMessage(responseHeader, ReleasableBytesReference.wrap(responseContent), () -> {});
 
         TransportHandshaker.HandshakeResponse response = new TransportHandshaker.HandshakeResponse(responseMessage.openOrGetStreamInput());
         assertThat(response.getResponseVersion()).isEqualTo(Version.V_6_0_0);
     }
 
+    /**
+     * Tests a handshake initiated by a newer node to an older node while the newer nodes minimum compatible version matches.
+     * <p/>
+     *  6.0.0 -> send handshake request -> 5.10.0       (5.10.0 accepts this as the minimum compatible version sent is 5.0.0)
+     *  5.10.0 -> send handshake response -> 6.0.0      (6.0.0 accepts this as the minimum compatible version defined is 5.0.0)
+     *
+     */
     @Test
     public void test_handshake_checks_minimum_compatible_version_if_normal_version_is_not_compatible() throws Exception {
-        // Create fake version which isn't compatible by its own, but it's minimum compatible version is compatible with the receiver.
-        Version remoteVersion = new Version(Version.V_4_8_4.internalId, Version.V_4_8_4.isSnapshot(), Version.V_4_8_4.luceneVersion) {
+        Version localVersion = new Version(Version.V_5_10_0.internalId, Version.V_5_10_0.isSnapshot(), Version.V_5_10_0.luceneVersion) {
             @Override
             public Version minimumCompatibilityVersion() {
-                return Version.fromString("5.0.0");
+                return Version.fromString("4.0.0");
             }
         };
```

Comment:
I wonder if we could change the default implementation of `minimumCompatibilityVersion` to be derived from the specific version of the `Version` instance (e.g. major - 1, 0, 0). I think that would allows us to use the regular Version constants and make sure the test really matches what we actually have in use.

Reply:
That would be the other way of solving this, yes. ES and CrateDB did this before, we changed it to be hardcoded by https://github.com/crate/crate/commit/3a7b21657ac1ef699770df76ea283a8e2ef5ad99#diff-347680c3d1b10c42c5c573aa1aaa876859db75afb106d4831e72735bd643e85d. Maybe non-intentionally? 

If we'd go this way, we could also change that the handshake request/response has an empty payload (the version is always transferred in the header, no need to transport the minimum version additionally).

I think I went with this approach to be less inversive. On the other hand, also this change needs to be backported.
I have no strong opinion here, shall we replace the hardcoded minimum version instead?

Reply:
> we changed it to be hardcoded by https://github.com/crate/crate/commit/3a7b21657ac1ef699770df76ea283a8e2ef5ad99#diff-347680c3d1b10c42c5c573aa1aaa876859db75afb106d4831e72735bd643e85d. Maybe non-intentionally?

I think it was because it didn't apply for us for 3 -> 4 and that was the most straightforward way to change it.

> I think I went with this approach to be less inversive. On the other hand, also this change needs to be backported.
> I have no strong opinion here, shall we replace the hardcoded minimum version instead?

Also no strong opinion and currently don't have a good intuition about the implications in terms of complexity, but if it isn't too much I'd prefer a version where tests don't diverge from the actual implementations.

Reply:
I've done this (compute minimalCompatibilityVersion) by https://github.com/crate/crate/pull/17264/commits/cfdd3e80d6e1dae2206bfea2d3f7eed8d79f058f, looks cleaner so I'd keep this.
Additionally, I bumped the `minimalCompatibilityVersion` of `6.x` to `5.10.1` as this will be the first one containing this patch and thus be able to support rolling upgrades to `6.x`.

 @mfussenegger could you please check again?

Reply:
btw. I've left the handshake request/response as it for BWC (I did not remove to stream the minimalCompatibilityVersion).